### PR TITLE
Do not measure variance for a conditional type extendsType

### LINKED ISSUE #31251: Comparison targets are reversed (regression)
The similar problem #30118 had been fixed, but appeared again.

**TypeScript Version:**  3.4.0-dev.20190504

**Search Terms:**

**Code**

https://github.com/falsandtru/spica/blob/v0.0.249/src/supervisor.legacy.ts#L56

**Expected behavior:**

pass

**Actual behavior:**

```
Argument of type 'this' is not assignable to parameter of type 'Supervisor<string, unknown, unknown, unknown>'.
  Type 'Supervisor<N, P, R, S>' is not assignable to type 'Supervisor<string, unknown, unknown, unknown>'.
    Type 'string' is not assignable to type 'N'.
      'string' is assignable to the constraint of type 'N', but 'N' could be instantiated with a different subtype of constraint 'string'.ts(2345)
```

**Playground Link:** 

**Related Issues:**

### PR DESCRIPTION
**Pull request by jack-williams**
Attempts to fix #31251

### CODE REVIEW

**Comment by falsandtru** [comment]
ping @weswigham

**Comment by jack-williams** [comment]
This PR does not address variance issues stemming from check types (see #31295 as an example). I wonder if it is best to try and fix them together?

**Comment by falsandtru** [comment]
ping @ahejlsberg @weswigham

**Comment by falsandtru** [comment]
ping @ahejlsberg @weswigham

**Comment by weswigham** [comment]
@jack-williams can you sync this? We'd want to test this across DT with perf in mind before merging it, which is our primary concern with using `Unmeasurable` and `Unreliable` more.

**Comment by jack-williams** [comment]
@weswigham done!

**Comment by weswigham** [comment]
@typescript-bot run dt
@typescript-bot test this

**Comment by falsandtru** [comment]
@jack-williams Can you resolve the errors?

@weswigham Can you run @typescript-bot pack this?

**Comment by weswigham** [comment]
Sure - @typescript-bot pack this

**Comment by falsandtru** [comment]
Thanks, but I can't install it: #31929

**Comment by falsandtru** [comment]
Thanks.

@jack-williams Comparison targets are still reversed:

```
Type 'typeof Book' does not satisfy the constraint 'typeof AnyModel'.
  Types of property 'reducer' are incompatible.
    Type '(action: RootAction, Book: ModelType<Book>) => void' is not assignable to type '(action: any, modelType: ModelType<any>, session: Pick<Pick<Session<any>, never> & Pick<{ [x: string]: ModelType<any>; }, never> & Pick<{ [x: string]: ModelType<any>; }, string | number | symbol>, string | ... 1 more ... | symbol>) => void'.
      Types of parameters 'Book' and 'modelType' are incompatible.
        Type 'ModelType<any>' is not assignable to type 'ModelType<Book>'.
          ...
```

It should be

```diff
-        Type 'ModelType<any>' is not assignable to type 'ModelType<Book>'.
+        Type 'ModelType<Book>' is not assignable to type 'ModelType<any>'.
```

**Comment by jack-williams** [comment]
@falsandtru can you give me the full example? This might be an issue with the check type, not the extends type.

**Comment by falsandtru** [comment]
See the 4352th line of https://typescript.visualstudio.com/TypeScript/_build/results?buildId=33063&view=logs&jobId=ae5f67bc-b508-52d0-18c7-3e0886a35c88&taskId=e2afb42f-d9e0-5ec8-c417-e22c6ef5223e&lineStart=4352&lineEnd=4353&colStart=1&colEnd=1

**Comment by jack-williams** [comment]
> It should be
`Type 'ModelType<Book>' is not assignable to type 'ModelType<any>'.`

Should it? The type appears in an argument position so I would expect them to be flipped when running `strictFunctionTypes`

I don't really have time to debug DT right now, but I wouldn't be surprised if this was correctly an error, something along the line of:
```ts
interface Foo<T,U> {
	x: T extends U ? number : string;
}
declare const a: Foo<any, boolean>;
let b: Foo<boolean, boolean> = a; // no error, but it should be
```

**Comment by falsandtru** [comment]
I referred to that line to explain the comparison order. Of course it should be no error if any type makes no errors.

**Comment by jack-williams** [comment]
@typescript-bot run dt because the old builds logs were cleared up.

**Comment by jack-williams** [comment]
I need to look into some of the changes to see whether they are reasonable, and more likely, what is the corresponding fix. Most of them seem to be cases where something is constrained with a type `Interface<any>` that previously passed on variance checks, but no longer does.

**Comment by weswigham** [comment]
I believe that. What we've been trying to do as of late is to fix the breaks on DT (via PRs to DT) such that we can get a clean run on the PR. It normally works out pretty well in cases like these where you can say "this was a possible way the old types could have been misused, TS with this new PR catches this, and this change preemptively fixes the package so it won't be broken by the next TS release".

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -13514,8 +13514,11 @@ namespace ts {
                         // Two conditional types 'T1 extends U1 ? X1 : Y1' and 'T2 extends U2 ? X2 : Y2' are related if
                         // one of T1 and T2 is related to the other, U1 and U2 are identical types, X1 is related to X2,
                         // and Y1 is related to Y2.
-                        if (isTypeIdenticalTo((<ConditionalType>source).extendsType, (<ConditionalType>target).extendsType) &&
-                            (isRelatedTo((<ConditionalType>source).checkType, (<ConditionalType>target).checkType) || isRelatedTo((<ConditionalType>target).checkType, (<ConditionalType>source).checkType))) {
+                        const u1 = instantiateType((<ConditionalType>source).extendsType, reportUnmeasurableMarkers);
+                        const u2 = instantiateType((<ConditionalType>target).extendsType, reportUnmeasurableMarkers);
```
Comment (weswigham) [inline src/compiler/checker.ts]: The source and target are identical sans type parameter instantiations during variance checking, so I've been wondering if we can get away with only instantiating one of the source or target.

Comment (jack-williams) [reply src/compiler/checker.ts]: Yeah that feels like it should work. At least, I don't see how replacing a type parameter with a sub/super type of itself could affect whether it gets instantiated.

**Comment by falsandtru** [comment]
Can you merge?

**Comment by weswigham** [comment]
I think it needs a refresh, a DT pass, an implementation of my last comment, probably a perf test just in case, and a once-over from @ahejlsberg once that's all done.

**Comment by falsandtru** [comment]
Okay.

**Comment by jack-williams** [comment]
@weswigham @falsandtru 

Yeah, sorry. The thing that's holding this up is trying to fix the breaks in DT caused by this. Context switching to a library I've never seen and trying to fix some obscure conditional type is time consuming, for me at least.

I'll at least make the instantiation simplification and re-run DT today or tomorrow. If anyone wants to help resolve DT breaks that would be much appreciated!

**Comment by jack-williams** [comment]
Tracking down one issues in react-select:

The following no longer checks:
```ts
export function manageState<T extends SelectBase<any>>(
  SelectComponent: T
): StateManager<GetOptionType<T>, T>;
```

fundamentally because of this conditional type in `react`:
```ts
    type WeakValidationMap<T> = {
        [K in keyof T]?: null extends T[K]
            ? Validator<T[K] | null | undefined>
            : undefined extends T[K]
            ? Validator<T[K] | null | undefined>
            : Validator<T[K]>
    };
```

Now we structurally compare this we end up with an identical relation check on for `T[K]` which fails because `any` is not identical to `GetOption<T>` where `T extends SelectBase<any>` and `type GetOptionType<T> = T extends SelectBase<infer OT> ? OT : never;`.

So we can fix this by doing this:
```ts
export function manageState<T extends SelectBase<U>, U = GetOptionType<T>>(
  SelectComponent: T
): StateManager<U, T>;
```
however we can also fix this by changing react:
```ts
    type WeakValidationMap<T> = {
        [K in keyof T]?: (x: T[K]) => void extends (x: null) => void
            ? Validator<T[K] | null | undefined>
            : (x: T[K]) => void extends (x: undefined) => void
            ? Validator<T[K] | null | undefined>
            : Validator<T[K]>
    };
```

I kind of prefer the `react` fix. What do you think @weswigham?

**Comment by jack-williams** [comment]
@typescript-bot run dt

**Comment by jack-williams** [comment]
@typescript-bot perf test this

**Comment by jack-williams** [comment]
A summary of the other breaks in DT.

`styled-components`: Again this is caused by the react `WeakValidationMap`. I can fix this by changing a type cast in the tests: `ThemedStyledComponentsModule<any>` -> `ThemedStyledComponentsModule<DefaultTheme>`

`wordpress__block-editor`: Alse caused by `WeakValidationMap`.

`redux-orm`: TODO

**Comment by weswigham** [comment]
The change to `WeakValidationMap` in react is probably fine, but likely deserves an explanation as to how wrapping the comparison in a signature changes the calculus, and why not doing so was incorrect.

**Comment by jack-williams** [comment]
@weswigham Fair enough. I guess given that check types are related bivariantly we are OK even if we consider those positions unmeasurable in the future?

**Comment by jack-williams** [comment]
@weswigham Sorry for the spam. 

Was looking into this more and I having your PR #30639 fixes all of the DT breaks. In all of the cases you end up with resolved conditional types on the LHS with `any`'s being related to unresolved conditional types on RHS (with type parameters).

**Comment by jack-williams** [comment]
@typescript-bot run dt

**Comment by jack-williams** [comment]
So, I flagrantly lied when I said that PR would fix everything. The breaks are as follows:

`ramda` breaks due to a dependency on `ts-toolbelt` and the `Curry` type, with the following:
```ts
// node_modules/ts-toolbelt/out/types/src/Function/Curry.d.ts(37,49):
// error TS2589: Type instantiation is excessively deep and possibly infinite.
```

`wordpress__block-editor` for similar issues as before. The break is good and I think easily fixed.

`redux-first-router` and `redux-orm` both run out of memory (I suspect the former depends on the latter).

I *think* that the oom and excessive instantiation errors are caused only by the change in #30639 and not by the composition of this feature.

Should we delay this issue until #30639 is resolved? I think making as few changes to DT as possible seems desirable.

**Comment by weswigham** [comment]
I'd believe #30639 needing to be a prerequisite to doing this with minimal breakage. I vaguely remember a break in ramda as being why I didn't merge it, so I suppose I should return to investigating that (though the oom is more cocerning, ofc).

**Comment by jack-williams** [comment]
Note to self: back out last commits adding conditional type assignability rule.

**Comment by pirix-gh** [comment]
@jack-williams would you mind packing this? thx

**Comment by jack-williams** [comment]
@pirix-gh I'll merge master and pack it; I think it would be good to make sure we also include the recursive alias changes. I'll try and get this done by Wednesday, if thats ok?

**Comment by jack-williams** [comment]
@typescript-bot pack this

**Comment by jack-williams** [comment]
@pirix-gh

**Comment by pirix-gh** [comment]
Thanks @jack-williams, it seems like ramda/Curry isn't breaking anymore. Is this your (close to) final version?

**Comment by jack-williams** [comment]
This version is sufficient to fix the associated problem, but it introduces some breaking changes that would need likely need #30639 to make palatable.

**Comment by falsandtru** [comment]
I found a regression on https://github.com/falsandtru/spica.

```
src/list.ts(46,13): Error TS2352: Conversion of type '(as extends [unknown, unknown, ...unknown[]] ? as : never)["length"] extends 1 ? Nil : Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' to type 'Cons<Split<as>[1]>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'Nil | Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' is not comparable to type 'Cons<Split<as>[1]>'.
    Type 'Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' is not comparable to type 'Cons<Split<as>[1]>'.
      Types of property 'tail' are incompatible.
        Type 'Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]["length"] extends 1 ? Nil : Cons<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>' is not comparable to type 'Split<as>[1]["length"] extends 1 ? Nil : Cons<Split<Split<as>[1]>[1]>'.
          Type 'Cons<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>' is not comparable to type 'Cons<Split<Split<as>[1]>[1]>'.
            Types of property 'tail' are incompatible.
              Type 'Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]["length"] extends 1 ? Nil : Cons<Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]>' is not comparable to type 'Split<Split<as>[1]>[1]["length"] extends 1 ? Nil : Cons<Split<Split<Split<as>[1]>[1]>[1]>'.
                Type 'Cons<Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]>' is not comparable to type 'Cons<Split<Split<Split<as>[1]>[1]>[1]>'.
                  Types of property 'tail' are incompatible.
                    Type 'Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]["length"] extends 1 ? Nil : Cons<Split<Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]>[1]>' is not comparable to type 'Split<Split<Split<as>[1]>[1]>[1]["length"] extends 1 ? Nil : Cons<Split<Split<Split<Split<as>[1]>[1]>[1]>[1]>'.
                      Type 'Cons<Split<Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]>[1]>' is not comparable to type 'Cons<Split<Split<Split<Split<as>[1]>[1]>[1]>[1]>'.      
                        The types returned by 'reverse()' are incompatible between these types.
                          Type 'Reverse<Split<Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]>[1]>' is not comparable to type 'Reverse<Split<Split<Split<Split<as>[1]>[1]>[1]>[1]>'.
```

**Comment by falsandtru** [comment]
The error come from `this.tail as Cons<Split<as>[1]>` but the error message of `this.tail as Cons<Split<AtLeast<2, unknown>>[1]>` would be more readable.

```
Conversion of type '(as extends [unknown, unknown, ...unknown[]] ? as : never)["length"] extends 1 ? Nil : Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' to type 'Cons<[unknown, ...unknown[]]>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'Nil | Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' is not comparable to type 'Cons<[unknown, ...unknown[]]>'.
    Type 'Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' is not comparable to type 'Cons<[unknown, ...unknown[]]>'.
      Types of property 'CONS' are incompatible.
        Type 'Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]' is not comparable to type '[unknown, ...unknown[]]'.
          Type 'Split<as>[1]' is not comparable to type '[unknown, ...unknown[]]'.
            Property '0' is missing in type 'unknown[]' but required in type '[unknown, ...unknown[]]'.ts(2352)
```

At the following place,

```
          Type 'Split<as>[1]' is not comparable to type '[unknown, ...unknown[]]'.
            Property '0' is missing in type 'unknown[]' but required in type '[unknown, ...unknown[]]'.ts(2352)
```

`Split<as>[1]` must be `[unknown, ...unknown[]]` but looks like the compiler assumes `unknown[]`.

**Comment by DanielRosenwasser** [comment]
@jack-williams pointed out this is still open on #36138. Did you get the chance to take a look at this @ahejlsberg?

**Comment by falsandtru** [comment]
https://github.com/microsoft/TypeScript/pull/31277#issuecomment-566969038 is missing. Can you pack again?

**Comment by jack-williams** [comment]
@typescript-bot pack this

**Comment by falsandtru** [comment]
Hm, the task has failed.

**Comment by jack-williams** [comment]
@typescript-bot pack this

**Comment by falsandtru** [comment]
Thanks. I cleaned the related type functions by removing unnecessary constraints but that made no effect. The current errors are as follows.

```
spica/src/list.ts(46,13): Error TS2352: Conversion of type '(as extends [unknown, unknown, ...unknown[]] ? as : never)["length"] extends 1 ? Nil : Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' to type 'Cons<Split<as>[1]>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'Nil | Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' is not comparable to type 'Cons<Split<as>[1]>'.
    Type 'Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' is not comparable to type 'Cons<Split<as>[1]>'.
      Types of property 'tail' are incompatible.
        Type 'Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]["length"] extends 1 ? Nil : Cons<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>' is not comparable to type 'Split<as>[1]["length"] extends 1 ? Nil : Cons<Split<Split<as>[1]>[1]>'.
          Type 'Cons<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>' is not comparable to type 'Cons<Split<Split<as>[1]>[1]>'.
            Types of property 'tail' are incompatible.
              Type 'Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]["length"] extends 1 ? Nil : Cons<Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]>' is not comparable to type 'Split<Split<as>[1]>[1]["length"] extends 1 ? Nil : Cons<Split<Split<Split<as>[1]>[1]>[1]>'.
                Type 'Cons<Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]>' is not comparable to type 'Cons<Split<Split<Split<as>[1]>[1]>[1]>'.
                  Types of property 'tail' are incompatible.
                    Type 'Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]["length"] extends 1 ? Nil : Cons<Split<Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]>[1]>' is not comparable to type 'Split<Split<Split<as>[1]>[1]>[1]["length"] extends 1 ? Nil : Cons<Split<Split<Split<Split<as>[1]>[1]>[1]>[1]>'. 
                      Type 'Cons<Split<Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]>[1]>' is not comparable to type 'Cons<Split<Split<Split<Split<as>[1]>[1]>[1]>[1]>'.
                        The types returned by 'reverse()' are incompatible between these types.
                          Type 'Reverse<Split<Split<Split<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>[1]>[1]>[1]>' is not comparable to type 'Reverse<Split<Split<Split<Split<as>[1]>[1]>[1]>[1]>'.
```

```
spica/src/list.ts(46,13): Error TS2352: Conversion of type '(as extends [unknown, unknown, ...unknown[]] ? as : never)["length"] extends 1 ? Nil : Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' to type 'Cons<[unknown, ...unknown[]]>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'Nil | Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' is not comparable to type 'Cons<[unknown, ...unknown[]]>'.
    Type 'Cons<Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]>' is not comparable to type 'Cons<[unknown, ...unknown[]]>'.
      Types of property 'CONS' are incompatible.
        Type 'Split<as extends [unknown, unknown, ...unknown[]] ? as : never>[1]' is not comparable to type '[unknown, ...unknown[]]'.
          Type 'Split<as>[1]' is not comparable to type '[unknown, ...unknown[]]'.
            Property '0' is missing in type 'unknown[]' but required in type '[unknown, ...unknown[]]'.
```

**Comment by sandersn** [comment]
It sounds like this PR is currently blocked on #30639. Is that a correct reading of the comment history @jack-williams?

**Comment by jack-williams** [comment]
@sandersn Blocked may be too strong, but I think it makes sense to merge #30639 first, if possible. This PR makes the compiler check more conditional types structurally, so having more relation rules seems good.

I believe #30639 would change _some_ of the breaks introduced by this PR, but I haven't checked DT in a while.

**Comment by weswigham** [comment]
@jack-williams #30639  got merged this release! Do we need to refresh/reevaluate breaks? I imagine so since the last update in here was march of last year.

**Comment by jack-williams** [comment]
@weswigham I've started trying to revive this. There is one new regression in the compiler tests which I have not been able to figure out yet, but I have been able to reduce to a small example:

```ts
type EqualsTest<T> = <A>() => A extends T ? 1 : 0;
type EqualsTest1<T> = <A>() => A extends T ? 1 : 0;

const x: EqualsTest<number> = undefined as any as EqualsTest<string>; // was error, now no error
const y: EqualsTest<number> = undefined as any as EqualsTest1<string>; // error;
```

Both used to fail, now only `y` does.

**Comment by weswigham** [comment]
Information loss during generic signature relating due to local type parameters erasure is I think likely at fault there, like I (tried to) explain in #43867.

**Comment by weswigham** [comment]
_Specifically_ the self-assignment case lacks the error because we have a "fast path" in signature comparison (in `signaturesRelatedTo`) for when we're relating two signatures from the same type in which _both_ copies of the signature have their type parameters erased to `any`, which in turn causes us to fully evaluate the conditional (to a union of both branches) when we do the structural comparison, which is why we fail to issue an error. Type erasure has another outstanding issue - #41067 - which I don't think would help here, since the wildcard is just a special `any`, so the conditionals would still incorrectly execute and eliminate the `extends` clause, but I think points to a possible solution - we need the "erased" type parameter to be assignable like an `any`, _but_ still prevent execution of conditionals, and thus look like a generic. That's neither `any`, nor the `wildcardType` right now, but I don't think it'd be too bad to clone one of the two and consider it "instantiable" in the context of constructing a conditional (and thus prevent its execution).

I have just put together a composition of this change, the change in #41067, and what I just explained, which seems to neatly tie all this together and get our conditionals-in-generic-signatures relationships to where I think it should be. I've put up #43887 which combines this with the other changes I think we need in place to get it across the finish line.

**Comment by jack-williams** [comment]
Thanks @weswigham, your description on the first thread was clear, I just didn't read it in enough detail. The `any` from the generic erasure causing problems makes sense.

I will close this up in favor of your new PR. Without your change this PR break `Equal` from `ts-toolbelt`, and likely other type heavy packages.

**Pull request closed without merging**